### PR TITLE
Wrote comparator and unit tests for comparator.

### PR DESCRIPTION
**Pull request by spiresheep**

### CODE REVIEW

File: source/comparator.ts [hunk 1]
```diff
@@ -10,6 +10,78 @@ export class Comparator {
    * @throws {TypeError} - The parameters can not be compared to one another.
    */
   public compareValues(left: any, right: any): number {
-    return 0;
+    if (left === right) {
```
Comment (kranar) [inline source/comparator.ts]: No space after ```if```.

File: source/comparator.ts [hunk 2]
```diff
@@ -10,6 +10,78 @@ export class Comparator {
...
+    }
+    if (left === undefined || right === undefined) {
+      if (left === undefined) {
+        return -1;
+      } else {
+        return 1;
+      }
+    }
+    if (left === null || right === null) {
+
```
Comment (kranar) [inline source/comparator.ts]: Remove blank line.

File: source/comparator.ts [hunk 3]
```diff
@@ -10,6 +10,78 @@ export class Comparator {
    * @throws {TypeError} - The parameters can not be compared to one another.
    */
   public compareValues(left: any, right: any): number {
-    return 0;
+    if (left === right) {
+      return 0;
+    }
+    if (left === undefined || right === undefined) {
```
Comment (kranar) [inline source/comparator.ts]: This ```if``` is redundant. Refactor lines 16 to 22.

File: source/comparator.ts [hunk 4]
```diff
@@ -10,6 +10,78 @@ export class Comparator {
...
+      if (left === undefined) {
+        return -1;
+      } else {
+        return 1;
+      }
+    }
+    if (left === null || right === null) {
```
Comment (kranar) [inline source/comparator.ts]: Similarly this check is redundant.

File: source/comparator.ts [hunk 5]
```diff
@@ -10,6 +10,78 @@ export class Comparator {
...
+      throw TypeError('The parameters can not be compared to one another');
+    }
+    if (typeof left === 'boolean') {
+      if (left) {
+        return 1;
+      } else {
+        return -1;
+      }
+    }
+    if (typeof left === 'number') {
```
Comment (kranar) [inline source/comparator.ts]: Use a ```switch``` instead of many ```if```s.

File: source/comparator.ts [hunk 6]
```diff
@@ -10,6 +10,78 @@ export class Comparator {
...
+      if (isFinite(left) && isFinite(right)) {
+        if (left > right) {
```
Comment (kranar) [inline source/comparator.ts]: ```return left - right;```

**Review by kranar — CHANGES_REQUESTED** [review]
Add unit tests.

File: source/comparator.ts [hunk 7]
```diff
@@ -10,78 +10,78 @@ export class Comparator {
...
-        return -1;
+    if(typeof left === 'object') {
+      if(!(left instanceof Date) || !(right instanceof Date)) {
+        throw TypeError('The parameters can not be compared to one another');
       }
     }
-    if (typeof left === 'number') {
-      if (isFinite(left) && isFinite(right)) {
-        if (left > right) {
+    switch (typeof left) {
```
Comment (kranar) [inline source/comparator.ts]: No space after keywords, ```switch(...)```.

File: source/comparator.ts [hunk 8]
```diff
@@ -10,78 +10,78 @@ export class Comparator {
...
         } else {
           return -1;
         }
-      } else {
-        if(left === -Infinity || right === -Infinity) {
-          if(right === -Infinity) {
+        break;
+      case 'number':
+        if(isFinite(left) && isFinite(right)) {
+          if(left > right) {
```
Comment (kranar) [inline source/comparator.ts]: Use ```return left - right;```

**Review by kranar — CHANGES_REQUESTED** [review]
Much better.

File: source/comparator.ts [hunk 9]
```diff
@@ -10,6 +10,73 @@ export class Comparator {
    * @throws {TypeError} - The parameters can not be compared to one another.
    */
   public compareValues(left: any, right: any): number {
-    return 0;
+    if(left === right) {
```
Comment (kranar) [inline source/comparator.ts:13]: Use ```if/else if``` to keep your code consistent throughout.

File: source/comparator.ts [hunk 10]
```diff
@@ -10,6 +10,73 @@ export class Comparator {
...
+            if(isNaN(right)) {
+              return 1;
+            } else {
+              return -1;
+            }
+          }
+        }
+        break;
+      case 'object':
+        if (left.valueOf() > right.valueOf()) {
```
Comment (kranar) [inline source/comparator.ts]: Remove space.

**Review by kranar — CHANGES_REQUESTED** [review]

File: source/comparator.ts [hunk 11]
```diff
@@ -10,6 +10,61 @@ export class Comparator {
    * @throws {TypeError} - The parameters can not be compared to one another.
    */
   public compareValues(left: any, right: any): number {
-    return 0;
+    if(left === right) {
+      return 0;
+    }
```
Comment (kranar) [inline source/comparator.ts]: Keep going with the if/else if.

File: source/comparator.ts [hunk 12]
```diff
@@ -10,6 +10,61 @@ export class Comparator {
...
+        if(left) {
+          return 1;
+        } else {
+          return -1;
+        }
+      case 'number':
+        if(isFinite(left) && isFinite(right)) {
+          return left - right;
+        } else {
+          if(right === -Infinity) {
```
Comment (kranar) [inline source/comparator.ts]: Use an ```else if``` instead of ```else { if```

File: source/comparator.ts [hunk 13]
```diff
@@ -10,6 +10,61 @@ export class Comparator {
...
+          } else if(left === Infinity) {
+            return 1;
+          } else if(right === Infinity) {
+            return -1;
+          } else if(isNaN(right)) {
+            return 1;
+          } else if(isNaN(left)) {
+            return -1;
+          }
+        }
```
Comment (kranar) [inline source/comparator.ts:52]: What happens if execution reaches line 57?

**Review by kranar — CHANGES_REQUESTED** [review]

File: source/comparator.ts (hunk 13 above)
Comment (spiresheep) [reply source/comparator.ts:52]: It means it was a number but it wasn't in any of the above cases.

Comment (kranar) [reply source/comparator.ts:52]: Good, but what do you return in that case?

Comment (spiresheep) [reply source/comparator.ts:52]: Throw an exception.

**Pull request merged**